cluster-api/pkg/apis/cluster/v1alpha1/machinedeployment_types.go

```diff
@@ -0,0 +1,269 @@
+	// be machines that are running but not yet available or machines
+	// that still have not been created.
+	// +optional
+	UnavailableReplicas int32 `json:"unavailableReplicas,omitempty" protobuf:"varint,5,opt,name=unavailableReplicas"`
+}
+
+// Validate checks that an instance of MachineDeployment is well formed
+func (MachineDeploymentValidationStrategy) Validate(ctx request.Context, obj runtime.Object) field.ErrorList {
+	md := obj.(*cluster.MachineDeployment)
+	log.Printf("Validating fields for MachineDeployment %s\n", o.Name)
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, ValidateMachineDeploymentStrategy(&md.Spec, fldPath.Child("spec"))...)
+	return allErrs
+}
+
+func ValidateMachineDeploymentSpec(spec *MachineDeploymentSpec, fldPath *field.Path) field.ErrorList {
+	allErrs = append(allErrs, metav1validation.ValidateLabelSelector(&spec.Selector, fldPath.Child("selector"))...)
+	if len(spec.Selector.MatchLabels)+len(spec.Selector.MatchExpressions) == 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "empty selector is not valid for MachineSet."))
+	}
+	selector, err := metav1.LabelSelectorAsSelector(&spec.Selector)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "invalid label selector."))
+	} else {
+		labels := labels.Set(spec.Template.Labels)
+		if !selector.Matches(labels) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("template", "metadata", "labels"), spec.Template.Labels, "`selector` does not match template `labels`"))
+		}
+	}
+
+	if *obj.Spec.Replicas < 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("replicas"), *obj.Spec.Replicas, "replicas can not be negative"))
+	}
+
+	allErrs = append(allErrs, ValidateMachineDeploymentStrategy(&spec.Strategy, fldPath.Child("strategy"))...)
+	return allErrs
+}
+
+func ValidateMachineDeploymentStrategy(strategy *MachineDeploymentStrategy, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	switch strategy.Type {
+	case common.RollingUpdateMachineDeploymentStrategyType:
+		allErrs = append(allErrs, ValidateMachineRollingUpdateDeployment(strategy.RollingUpdate, fldPath.Child("rollingUpdate"))...)
+	}
+	return allErrs
+}
+
+func ValidateMachineRollingUpdateDeployment(rollingUpdate *MachineRollingUpdateDeployment, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, ValidatePositiveIntOrPercent(rollingUpdate.MaxUnavailable, fldPath.Child("maxUnavailable"))...)
+	allErrs = append(allErrs, ValidatePositiveIntOrPercent(rollingUpdate.MaxSurge, fldPath.Child("maxSurge"))...)
+	if getIntOrPercentValue(rollingUpdate.MaxUnavailable) == 0 && getIntOrPercentValue(rollingUpdate.MaxSurge) == 0 {
+		// Both MaxSurge and MaxUnavailable cannot be zero.
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("maxUnavailable"), rollingUpdate.MaxUnavailable, "may not be 0 when `maxSurge` is 0"))
```

Comment:
This message might be confusing if they used percentages. Is there a better message? Maybe not.

Reply:
Not sure, I just copied the error message from `Deployment`